### PR TITLE
[infra] Support GCC as compiler for Taichi developers (experimental)

### PR DESCRIPTION
**Pull request by archibate**
Related issue = #

[[Click here for the format server]](http://kun.csail.mit.edu:31415/)

----
IMO, GCC vs Clang ~= OpenGL vs CUDA: despite Clang is more fit with LLVM, but GCC is distributed on almost all *nix machines, we may also add mingw64 support later instead of the ad-hoc MSVC which is embed into that gaint IDE..

It also provides many useful warnings (much more than Clang) that might help us debug our C++ program.

### CODE REVIEW

File: cmake/TaichiCXXFlags.cmake [hunk 1]
```diff
@@ -25,7 +25,11 @@ else()
     if ("${CMAKE_CXX_COMPILER_ID}" MATCHES "Clang")
         message("Clang compiler detected. Using std=c++17.")
         set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -std=c++17 -fsized-deallocation")
+    elseif ("${CMAKE_CXX_COMPILER_ID}" MATCHES "GNU")
+        message("GNU compiler detected. Using std=c++17.")
+        set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -std=c++17 -fsized-deallocation -Wno-class-memaccess -Wno-comment -Wno-sign-compare")
     else()
+        message("Invalid compiler ${CMAKE_CXX_COMPILER_ID} detected.")
         message(FATAL_ERROR "clang and MSVC are the only supported compilers for Taichi compiler development. Consider using 'cmake -DCMAKE_CXX_COMPILER=clang' if you are on Linux")
```
Comment (archibate) [inline cmake/TaichiCXXFlags.cmake:33]: Will update this message and doc when GCC is battle-tested.

**Comment by TH3CHARLie** [comment]
It would be great if GCC is supported.

> 
> It also provides many useful warnings (much more than Clang) that might help us debug our C++ program.

For this, I think maybe adding another build(possibly: master build, ubuntu, GCC and 3.7) to monitor and also show the effect of GCC. Since GCC support is now experimental, the build can be set to allow failure. It can happen within this PR or in a separate one, either way, having this would enable all taichi developers to monitor GCC progress and have a better context on it.

**Comment by archibate** [comment]
> It's worth mentioning that if someone uses the LLVM backend for development, `clang` is still needed to compile the runtime bitcode.

Yes, but note that we don't need `clang` if one only do development on `cc`, `opengl` or `metal`, which are s2s backends. I didn't update the doc yet since it's experimental for now.

> I guess you are trying to use the `cc` backend with gcc (and therefore you don't need the LLVM backend?)

Yes, that's my point: to allow contributors build and test on `cc` and `opengl` without LLVM - which is our configured one. This make contributions to C++ part easier. To do so, we need something like `TI_WITH_LLVM:BOOL=OFF`, would you help me do this iapr when you've time?

> It's nice that we can compile Taichi with `gcc`. Feel free to merge if it works on your end.

Yes, that's cool and many examples are functional! But some tests involving advanced features failed, maybe it's `gcc -O3`'s fault, I'll try to fix them before merge, or iapr?

**Comment by yuanming-hu** [comment]
Thanks. All sounds good to me.

> To do so, we need something like `TI_WITH_LLVM:BOOL=OFF`, would you help me do this iapr when you've time?

I think it's a good idea to add that flag. I do it when I have time, but I'm probably slow given my recent schedule. It can be a nice newcomer issue. Maybe we should open an issue and let new contributors help with that?

> Yes, that's cool and many examples are functional! But some tests involving advanced features failed, maybe it's `gcc -O3`'s fault, I'll try to fix them before merge, or iapr?

It's great that many examples are working! I'd suggest we do that in a different PR. Small PRs are always good.

**Pull request merged**